python/lcextension/ext.py

```diff
@@ -158,4 +161,131 @@ def logCritical( self, msg ):
+        hive = limacharlie.Hive(manager, "cloud_sensor", manager._oid)
+        try:
+            hive.delete(self._name)
+        except Exception as e:
+                if not (isinstance(e, limacharlie.LcApiException) and "RECORD_NOT_FOUND" in str(e)):
+                    raise limacharlie.LcApiException(f"failed hive delete: {e}")
+        
+        install_key_desc = self.get_extension_adapter_installation_key_desc()
+        try:
+            install_keys = manager.get_installation_keys()
+        except Exception as e:
+            raise limacharlie.LcApiException(f"failed to list installation keys: {e}")
+        
+        private_tag = self.get_extension_private_tag()
+
+        for org_id, keys in install_keys.items():
+            for key_id, key in keys.items():
+                if key['desc'] != install_key_desc:
+                    continue
+
+                tags = [tag.strip() for tag in key['tags'].split(',') if tag.strip()]
+                is_tag_found = private_tag in tags
+                if not is_tag_found:
+                    continue
+
+                try:
+                    manager.delete_installation_key(key['iid'])
+                except Exception as e:
+                    if not (isinstance(e, limacharlie.LcApiException) and "RECORD_NOT_FOUND" in str(e)):
+                        raise limacharlie.LcApiException(f"Failed to delete installation key {key['iid']: [e]}")
+
+
+    def generate_webhook_secret_for_org(self, oid):
+        # This generates a secret value deterministically from
+        # the OID so that we can easily know the webhook to
+        # hit without having to query LC. The WEBHOOK_SECRET
+        # needs to remain secret to avoid someone possibly
+        # sending their own data to users.
+        h = hashlib.sha256()
+        h.update(self._secret.encode())
+        h.update(oid.encode())
+        return h.hexdigest()
+    
+    def get_extension_adapter_installation_key_desc(self):
+        # Returns a description string for the extension adapter installation key
+        return f"ext {self._name} webhook adapter"
+
+    def get_adapter_client(self, manager):
+        # try to get the client if it already exists
+        with self.wh_clients_lock:
+            client = self.wh_clients.get(manager._oid)
+
+        if client:
+            return client
+
+        # Create a new client if it doesn't exist
+        try:
+            new_client = limacharlie.WebhookSender(manager, self._name, self.generate_webhook_secret_for_org(manager._oid))
+        except Exception as e:
+            raise Exception(f"failed to create webhook sender client: {e} ")
+
+        with self.wh_clients_lock:
+            # double check if client was added before add
+            client = self.wh_clients.get(manager._oid)
+            if client:
+                new_client.close()
+                return client
+
+            self.wh_clients[manager._oid] = new_client
```

Comment:
```suggestion
        with self.wh_clients_lock:
            client = self.wh_clients.get(manager._oid)

            if client:
                return client

            # Create a new client if it doesn't exist
            try:
                new_client = limacharlie.WebhookSender(manager, self._name, self.generate_webhook_secret_for_org(manager._oid))
            except Exception as e:
                raise Exception(f"failed to create webhook sender client: {e} ")

            # double check if client was added before add
            client = self.wh_clients.get(manager._oid)
            if client:
                new_client.close()
                return client

            self.wh_clients[manager._oid] = new_client
```

I think you want the whole thing to be under the same lock to avoid a race condition while creating the object. Not a huge deal though.

Reply:
good catch logic updated and removed second check since we are obtaining lock at beginning of call